Unable to compile under Ubuntu Linux due to dependency on `WSA`

Hey @Aniseto,

Here's the error output from the compiler:

```
#####
# Building NosoWalletCPP
#####
[ 25%] Building CXX object CMakeFiles/NosoWalletCPP.dir/NosoWalletCPP.cpp.o
In file included from /home/gcarreno/Programming/Noso/gcarreno-NosoWalletCPP/NosoWalletCPP.cpp:6:
/home/gcarreno/Programming/Noso/gcarreno-NosoWalletCPP/Communication.h: In function ‘std::string SendStringToNode(const std::string&, int, const std::string&)’:
/home/gcarreno/Programming/Noso/gcarreno-NosoWalletCPP/Communication.h:10:5: error: ‘WSADATA’ was not declared in this scope
   10 |     WSADATA wsaData; // Wsdata Variable
      |     ^~~~~~~
/home/gcarreno/Programming/Noso/gcarreno-NosoWalletCPP/Communication.h:11:20: error: ‘MAKEWORD’ was not declared in this scope
   11 |     if (WSAStartup(MAKEWORD(2, 2), &wsaData) != 0) {
      |                    ^~~~~~~~
/home/gcarreno/Programming/Noso/gcarreno-NosoWalletCPP/Communication.h:11:37: error: ‘wsaData’ was not declared in this scope
   11 |     if (WSAStartup(MAKEWORD(2, 2), &wsaData) != 0) {
      |                                     ^~~~~~~
/home/gcarreno/Programming/Noso/gcarreno-NosoWalletCPP/Communication.h:11:9: error: ‘WSAStartup’ was not declared in this scope
   11 |     if (WSAStartup(MAKEWORD(2, 2), &wsaData) != 0) {
      |         ^~~~~~~~~~
/home/gcarreno/Programming/Noso/gcarreno-NosoWalletCPP/Communication.h:17:5: error: ‘SOCKET’ was not declared in this scope
   17 |     SOCKET socketCliente = socket(AF_INET, SOCK_STREAM, 0);
      |     ^~~~~~
/home/gcarreno/Programming/Noso/gcarreno-NosoWalletCPP/Communication.h:18:9: error: ‘socketCliente’ was not declared in this scope
   18 |     if (socketCliente == INVALID_SOCKET) {
      |         ^~~~~~~~~~~~~
/home/gcarreno/Programming/Noso/gcarreno-NosoWalletCPP/Communication.h:18:26: error: ‘INVALID_SOCKET’ was not declared in this scope
   18 |     if (socketCliente == INVALID_SOCKET) {
      |                          ^~~~~~~~~~~~~~
/home/gcarreno/Programming/Noso/gcarreno-NosoWalletCPP/Communication.h:20:9: error: ‘WSACleanup’ was not declared in this scope
   20 |         WSACleanup();
      |         ^~~~~~~~~~
/home/gcarreno/Programming/Noso/gcarreno-NosoWalletCPP/Communication.h:30:21: error: ‘socketCliente’ was not declared in this scope
   30 |         closesocket(socketCliente);
      |                     ^~~~~~~~~~~~~
/home/gcarreno/Programming/Noso/gcarreno-NosoWalletCPP/Communication.h:30:9: error: ‘closesocket’ was not declared in this scope
   30 |         closesocket(socketCliente);
      |         ^~~~~~~~~~~
/home/gcarreno/Programming/Noso/gcarreno-NosoWalletCPP/Communication.h:31:9: error: ‘WSACleanup’ was not declared in this scope
   31 |         WSACleanup();
      |         ^~~~~~~~~~
/home/gcarreno/Programming/Noso/gcarreno-NosoWalletCPP/Communication.h:36:17: error: ‘socketCliente’ was not declared in this scope
   36 |     if (connect(socketCliente, reinterpret_cast<struct sockaddr*>(&direccionServidor), sizeof(direccionServidor)) == SOCKET_ERROR) {
      |                 ^~~~~~~~~~~~~
/home/gcarreno/Programming/Noso/gcarreno-NosoWalletCPP/Communication.h:36:118: error: ‘SOCKET_ERROR’ was not declared in this scope; did you mean ‘SO_ERROR’?
   36 |     if (connect(socketCliente, reinterpret_cast<struct sockaddr*>(&direccionServidor), sizeof(direccionServidor)) == SOCKET_ERROR) {
      |                                                                                                                      ^~~~~~~~~~~~
      |                                                                                                                      SO_ERROR
/home/gcarreno/Programming/Noso/gcarreno-NosoWalletCPP/Communication.h:38:9: error: ‘closesocket’ was not declared in this scope
   38 |         closesocket(socketCliente);
      |         ^~~~~~~~~~~
/home/gcarreno/Programming/Noso/gcarreno-NosoWalletCPP/Communication.h:39:9: error: ‘WSACleanup’ was not declared in this scope
   39 |         WSACleanup();
      |         ^~~~~~~~~~
/home/gcarreno/Programming/Noso/gcarreno-NosoWalletCPP/Communication.h:44:14: error: ‘socketCliente’ was not declared in this scope
   44 |     if (send(socketCliente, message.c_str(), message.size(), 0) == SOCKET_ERROR) {
      |              ^~~~~~~~~~~~~
/home/gcarreno/Programming/Noso/gcarreno-NosoWalletCPP/Communication.h:44:68: error: ‘SOCKET_ERROR’ was not declared in this scope; did you mean ‘SO_ERROR’?
   44 |     if (send(socketCliente, message.c_str(), message.size(), 0) == SOCKET_ERROR) {
      |                                                                    ^~~~~~~~~~~~
      |                                                                    SO_ERROR
/home/gcarreno/Programming/Noso/gcarreno-NosoWalletCPP/Communication.h:46:9: error: ‘closesocket’ was not declared in this scope
   46 |         closesocket(socketCliente);
      |         ^~~~~~~~~~~
/home/gcarreno/Programming/Noso/gcarreno-NosoWalletCPP/Communication.h:47:9: error: ‘WSACleanup’ was not declared in this scope
   47 |         WSACleanup();
      |         ^~~~~~~~~~
/home/gcarreno/Programming/Noso/gcarreno-NosoWalletCPP/Communication.h:60:39: error: ‘socketCliente’ was not declared in this scope
   60 |         while ((bytesRecibidos = recv(socketCliente, buffer, sizeof(buffer) - 1, 0)) > 0)
      |                                       ^~~~~~~~~~~~~
/home/gcarreno/Programming/Noso/gcarreno-NosoWalletCPP/Communication.h:74:35: error: ‘socketCliente’ was not declared in this scope
   74 |     while ((bytesRecibidos = recv(socketCliente, buffer, sizeof(buffer) - 1, 0)) > 0) {
      |                                   ^~~~~~~~~~~~~
/home/gcarreno/Programming/Noso/gcarreno-NosoWalletCPP/Communication.h:80:17: error: ‘socketCliente’ was not declared in this scope
   80 |     closesocket(socketCliente);
      |                 ^~~~~~~~~~~~~
/home/gcarreno/Programming/Noso/gcarreno-NosoWalletCPP/Communication.h:80:5: error: ‘closesocket’ was not declared in this scope
   80 |     closesocket(socketCliente);
      |     ^~~~~~~~~~~
/home/gcarreno/Programming/Noso/gcarreno-NosoWalletCPP/Communication.h:81:5: error: ‘WSACleanup’ was not declared in this scope
   81 |     WSACleanup();
      |     ^~~~~~~~~~
gmake[2]: *** [CMakeFiles/NosoWalletCPP.dir/build.make:104: CMakeFiles/NosoWalletCPP.dir/NosoWalletCPP.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/NosoWalletCPP.dir/all] Error 2
gmake: *** [Makefile:91: all] Error 2
```

Hope this helps!!

Cheers,
Gus

--- a/.gitignore
+++ b/.gitignore
@@ -361,3 +361,6 @@ MigrationBackup/
 
 # Fody - auto-generated XML schema
 FodyWeavers.xsd
+
+# CMake build folder
+/build/

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,0 +1,31 @@
+cmake_minimum_required(VERSION 3.20)
+
+# specify the C++ standard
+## this should be equal to -std=c++20
+SET (CMAKE_CXX_STANDARD             20)
+SET (CMAKE_CXX_STANDARD_REQUIRED    True)
+
+# set the project name
+project(
+  NosoWalletCPP
+  VERSION 0.1
+  LANGUAGES CXX
+  HOMEPAGE_URL "https://github.com/Aniseto/NosoWalletCPP"
+)
+
+# Find required packages
+## WxWidgets
+find_package(wxWidgets REQUIRED gl core base OPTIONAL_COMPONENTS net)
+include(${wxWidgets_USE_FILE})
+
+# add the executable
+add_executable(NosoWalletCPP
+  App.cpp
+  DataStructures.cpp
+  NosoWalletCPP.cpp
+)
+
+# add the libs to the linker
+target_link_libraries(NosoWalletCPP PRIVATE
+  ${wxWidgets_LIBRARIES}
+)

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -28,4 +28,5 @@ add_executable(NosoWalletCPP
 # add the libs to the linker
 target_link_libraries(NosoWalletCPP PRIVATE
   ${wxWidgets_LIBRARIES}
+  crypto++
 )

--- a/Communication.h
+++ b/Communication.h
@@ -1,17 +1,27 @@
-#include <string>
 #pragma once
-#include <boost\asio.hpp> //Boost Library //vcpkg install boost:x64-windows-static https://www.boost.org/
+#include <string>
+#include <boost/asio.hpp> //Boost Library //vcpkg install boost:x64-windows-static https://www.boost.org/
 #include <iostream>
 #include <fstream>
 #include <vector>
 
+#ifndef _WIN32
+#include<sys/socket.h>
+#define SOCKET int
+#define INVALID_SOCKET -1
+#define WSACleanup() void()
+#define closesocket(s) close(s)
+#define SOCKET_ERROR -1
+#endif
 
 std::string SendStringToNode(const std::string& ip, int port, const std::string& message) {
+#ifdef _WIN32
     WSADATA wsaData; // Wsdata Variable
     if (WSAStartup(MAKEWORD(2, 2), &wsaData) != 0) {
         std::cerr << "Error Initializing Winsock." << std::endl;
         return "Error Initializing Winsock.";
     }
+#endif
 
     // Create TCP socket
     SOCKET socketCliente = socket(AF_INET, SOCK_STREAM, 0);

--- a/README.md
+++ b/README.md
@@ -6,9 +6,9 @@ c++ Based Noso Wallet
 
 Libraries:
 
--WxWidgets(https://www.wxwidgets.org/)
--Boost(https://www.boost.org/)
--CryptoPP(https://www.cryptopp.com/)
+- WxWidgets(https://www.wxwidgets.org/)
+- Boost(https://www.boost.org/)
+- CryptoPP(https://www.cryptopp.com/)
 
 Follow how to install those libraries from respective Websites, if using Visual Studio then you can use VCPKG(https://vcpkg.io/en/) package manager to compile and install libraries.
 

--- a/README.md
+++ b/README.md
@@ -14,4 +14,18 @@ Follow how to install those libraries from respective Websites, if using Visual 
 
 Remember to add respective Headers and Libs on your Visual studio solution Properties.
 
+## Build with CMake
 
+We are providing the `CMakeLists.txt` file to allow you to build this project with CMake
+
+### Building under Linux
+
+In order to compile this project under Linux you'll need to have the above libraries installed in your system
+
+**For Ubuntu**
+
+```console
+$ sudo apt install -y libboost-all-dev libcrypto++-dev libwxgtk3.2-dev
+```
+
+We are also providing the file `build_with_cmake.sh` as a quick way of invoking all the necessary commands in order to build with CMake.

--- a/README.md
+++ b/README.md
@@ -16,11 +16,11 @@ Remember to add respective Headers and Libs on your Visual studio solution Prope
 
 ## Build with CMake
 
-We are providing the `CMakeLists.txt` file to allow you to build this project with CMake
+We are providing the `CMakeLists.txt` file to allow you to build this project with CMake.
 
 ### Building under Linux
 
-In order to compile this project under Linux you'll need to have the above libraries installed in your system
+In order to compile this project under Linux you'll need to have the above libraries installed in your system.
 
 **For Ubuntu**
 

--- a/build_with_cmake.sh
+++ b/build_with_cmake.sh
@@ -1,0 +1,35 @@
+#!/bin/bash
+
+SOURCE_PATH=/home/gcarreno/Programming/Noso/NosoWalletCPP
+BUILD_PATH=${SOURCE_PATH}/build
+
+echo "###################################"
+echo "# IMPORTANT"
+echo "# In order to compile this project you'll need the following libs installed:"
+echo "# boost: sudo apt install -y libboost-all-dev"
+echo "# cryptopp: sudo apt install -y libcrypto++-dev"
+echo "# wxwidgets: sudo apt install -y libwxgtk3.2-dev"
+echo "#"
+echo "###################################"
+
+echo "Press <ENTER> to continue or <CTRL-C> to abort"
+read
+
+if [ ! -d "${BUILD_PATH}" ]
+then
+  mkdir "${BUILD_PATH}"
+fi
+
+echo "#####"
+echo "# Creating CMake build files"
+echo "#####"
+cmake -B "${BUILD_PATH}" -S "${SOURCE_PATH}"
+
+if [ $? -eq 0 ]
+then
+  echo
+  echo "#####"
+  echo "# Building NosoWalletCPP"
+  echo "#####"
+  cmake --build ${BUILD_PATH}
+fi

--- a/build_with_cmake.sh
+++ b/build_with_cmake.sh
@@ -1,6 +1,6 @@
 #!/bin/bash
 
-SOURCE_PATH=/home/gcarreno/Programming/Noso/NosoWalletCPP
+SOURCE_PATH=.
 BUILD_PATH=${SOURCE_PATH}/build
 
 echo "###################################"

